{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "heated-strip",
   "metadata": {},
   "source": [
    "# A. Input data-sets\n",
    "\n",
    "pyPARAGON requires two inputs: \n",
    "\n",
    "    i) A reference network which is composed of already known interactions, \n",
    "    ii) Seed nodes which can come from in omics datasets, or be frequent mutations in disases. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "saving-teach",
   "metadata": {},
   "source": [
    "### Preparing a reference network\n",
    "\n",
    "A reference network is introduced to pyPARAGON as a networkx object. Reference networks can include nan-nan interactions, causing troubles in calculations. Thus, we exclude nan-nan interactions in reference networks. Additionally, pyPARAGON does not consider self-interactions in graphlets. Therefore, we also eliminate self interactions to get rid of computational problems. Except for nan-nan and self-interactions, we may not use any filtration or additional preprocessing steps. pyPARAGON is able to eliminate the interactions with low confidence scores thanks to the flux calculation. Thus, we suggest not applying a confidence threshold value for the filtration. However, the user can eliminate the edges with low confidence scores. As an example, we excluded edges with an edge score less than 0.2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "mineral-separate",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd, networkx as nx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "handy-retailer",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_1</th>\n",
       "      <th>Gene_2</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ALDH1A1</td>\n",
       "      <td>ALDH1A1</td>\n",
       "      <td>0.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ITGA7</td>\n",
       "      <td>CHRNA1</td>\n",
       "      <td>0.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PPP1R9A</td>\n",
       "      <td>ACTG1</td>\n",
       "      <td>0.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRGN</td>\n",
       "      <td>CD44</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRB7</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>0.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783177</th>\n",
       "      <td>STIL</td>\n",
       "      <td>CDK5RAP2</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783178</th>\n",
       "      <td>SIRT7</td>\n",
       "      <td>SNORA10</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783179</th>\n",
       "      <td>BMP2</td>\n",
       "      <td>SNX18</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783180</th>\n",
       "      <td>C1QBP</td>\n",
       "      <td>MRPL50</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783181</th>\n",
       "      <td>CCNF</td>\n",
       "      <td>LMNB2</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>783182 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene_1    Gene_2  Score\n",
       "0       ALDH1A1   ALDH1A1   0.76\n",
       "1         ITGA7    CHRNA1   0.73\n",
       "2       PPP1R9A     ACTG1   0.65\n",
       "3          SRGN      CD44   0.63\n",
       "4          GRB7     ERBB2   0.90\n",
       "...         ...       ...    ...\n",
       "783177     STIL  CDK5RAP2   0.63\n",
       "783178    SIRT7   SNORA10   0.63\n",
       "783179     BMP2     SNX18   0.63\n",
       "783180    C1QBP    MRPL50   0.63\n",
       "783181     CCNF     LMNB2   0.63\n",
       "\n",
       "[783182 rows x 3 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HIPPIE_df=pd.read_csv(\"../Source/Interactomes/HIPPIE_v2_3.tab\", keep_default_na=False, sep=\"\\t\") # excluding rows with nan value\n",
    "HIPPIE_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "postal-subsection",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_1</th>\n",
       "      <th>Gene_2</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ALDH1A1</td>\n",
       "      <td>ALDH1A1</td>\n",
       "      <td>0.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ITGA7</td>\n",
       "      <td>CHRNA1</td>\n",
       "      <td>0.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PPP1R9A</td>\n",
       "      <td>ACTG1</td>\n",
       "      <td>0.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRGN</td>\n",
       "      <td>CD44</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRB7</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>0.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783177</th>\n",
       "      <td>STIL</td>\n",
       "      <td>CDK5RAP2</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783178</th>\n",
       "      <td>SIRT7</td>\n",
       "      <td>SNORA10</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783179</th>\n",
       "      <td>BMP2</td>\n",
       "      <td>SNX18</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783180</th>\n",
       "      <td>C1QBP</td>\n",
       "      <td>MRPL50</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783181</th>\n",
       "      <td>CCNF</td>\n",
       "      <td>LMNB2</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>777762 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene_1    Gene_2  Score\n",
       "0       ALDH1A1   ALDH1A1   0.76\n",
       "1         ITGA7    CHRNA1   0.73\n",
       "2       PPP1R9A     ACTG1   0.65\n",
       "3          SRGN      CD44   0.63\n",
       "4          GRB7     ERBB2   0.90\n",
       "...         ...       ...    ...\n",
       "783177     STIL  CDK5RAP2   0.63\n",
       "783178    SIRT7   SNORA10   0.63\n",
       "783179     BMP2     SNX18   0.63\n",
       "783180    C1QBP    MRPL50   0.63\n",
       "783181     CCNF     LMNB2   0.63\n",
       "\n",
       "[777762 rows x 3 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HIPPIE_df=HIPPIE_df[HIPPIE_df[\"Score\"]>0.2]\n",
    "HIPPIE_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "tested-april",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of edges in HIPPIE network 769054\n",
      "The number of nodes in HIPPIE network 19433\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Networkx object and removing self interactions\n",
    "HIPPIE_nx=nx.from_pandas_edgelist(HIPPIE_df,\"Gene_1\",\"Gene_2\",edge_attr=\"Score\")\n",
    "self_interactions=list(nx.selfloop_edges(HIPPIE_nx))\n",
    "\n",
    "\n",
    "for u, v in self_interactions:\n",
    "    if isinstance(u, float):\n",
    "        print(u)\n",
    "    else:\n",
    "        HIPPIE_nx.remove_edge(u,v)\n",
    "    \n",
    "HIPPIE_df=nx.to_pandas_edgelist(HIPPIE_nx)\n",
    "print(\"The number of edges in HIPPIE network\",HIPPIE_nx.number_of_edges())\n",
    "print(\"The number of nodes in HIPPIE network\",HIPPIE_nx.number_of_nodes())\n",
    "\n",
    "\n",
    "edges=tuple(zip(HIPPIE_df.source,HIPPIE_df.target))\n",
    "for edge in edges:\n",
    "    if edge not in HIPPIE_nx.edges:\n",
    "        print(\"check the edge\",edge)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "expired-tours",
   "metadata": {},
   "source": [
    "###  Preparing seed nodes\n",
    "\n",
    "Seed nodes are known or detected context-specific knowledge. They can come from omics datasets, multi-omics datasets, or well-known, frequent mutations in diseases. Score your seeds as 1 if you do not generate initial node weights from raw data.\n",
    "\n",
    "During multi-omics data integration, the user can follow two strategies with pyPARAGON. In the first strategy,  your hits in omics datasets can be transformed into nodes represented in the reference network. Identifying possible transcription factors which target differential hits in transcriptomics or bind to the genome can be useful to carry transcriptomics and genomics knowledge into the interactomes.  With the second use option of pyPARAGON in multi-omic datasets, you may extend the reference network by merging various types of networks, such as regulator networks, interactomes, metabolic networks, etc. In this case, you can directly use multi-omics hits, which are represented in the reference network. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "chemical-fifty",
   "metadata": {},
   "outputs": [],
   "source": [
    "initial_nodes_df=pd.read_csv(f'../Source/Netpath/Sampling_0_5/AndrogenReceptor_05A/AndrogenReceptor_05A_var_0.nodes',sep=\"\\t\")\n",
    "\n",
    "initial_nodes=initial_nodes_df.name.to_list()\n",
    "initial_weights=[1 for _ in initial_nodes]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "featured-offense",
   "metadata": {},
   "source": [
    "## B) Constructing Graphlet Guided Network \n",
    "\n",
    "During the construction of a graphlet-guided network (GGN), pyPARAGON does not consider the confidence scores of edges in parsing the graphlets. The advanced usage of pyPARAGON in the construction of GGN is detailed in the Graphlet_Guided_Network_Construction, covering reference network permutations, graphlet frequencies, and graphlet motif selections. In this script, we recruit four-node graphlets, such as Graphlet5, Graphlet6, Graphlet7, and Graphlet8. During graphlet generation, we derived four-node graphlets from three-node graphlets. Also you can directly select four-node graphlets in pyPARAGON to overcome the computational cost. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "northern-threshold",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Paragon import GraphletGuidance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "tight-coffee",
   "metadata": {},
   "outputs": [],
   "source": [
    "Graphlet_list=['Graphlets5', 'Graphlets6', 'Graphlets7', 'Graphlets8']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "involved-briefing",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "GRF_lite=GraphletGuidance(HIPPIE_nx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "missing-tuning",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "82 of 82 input nodes have been found in the given network\n",
      "\n"
     ]
    }
   ],
   "source": [
    "GGN_nx=GRF_lite.construct_GGN(initial_nodes,Graphlets=Graphlet_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "designing-world",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4137"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GGN_nx.number_of_nodes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "numeric-ethnic",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSN</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSN</td>\n",
       "      <td>SRC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GSN</td>\n",
       "      <td>PCNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GSN</td>\n",
       "      <td>PLEKHA7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GSN</td>\n",
       "      <td>PRPF8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20805</th>\n",
       "      <td>IL6ST</td>\n",
       "      <td>RHOJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20806</th>\n",
       "      <td>IL6ST</td>\n",
       "      <td>RHOQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20807</th>\n",
       "      <td>IL6ST</td>\n",
       "      <td>RHOC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20808</th>\n",
       "      <td>IL6ST</td>\n",
       "      <td>SOCS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20809</th>\n",
       "      <td>IL6ST</td>\n",
       "      <td>SGTA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20810 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      source   target\n",
       "0        GSN         \n",
       "1        GSN      SRC\n",
       "2        GSN     PCNA\n",
       "3        GSN  PLEKHA7\n",
       "4        GSN    PRPF8\n",
       "...      ...      ...\n",
       "20805  IL6ST     RHOJ\n",
       "20806  IL6ST     RHOQ\n",
       "20807  IL6ST     RHOC\n",
       "20808  IL6ST    SOCS3\n",
       "20809  IL6ST     SGTA\n",
       "\n",
       "[20810 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GGN_df=nx.to_pandas_edgelist(GGN_nx)\n",
    "GGN_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "superb-exclusion",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GRF_lite.write_guided_graphlet_network(f'../Outputs/AndrogenReceptor_GGN')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "tired-albuquerque",
   "metadata": {},
   "source": [
    "## C) PageRank flux and context-specific network inference"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fluid-providence",
   "metadata": {},
   "source": [
    "pyPARAGON, before infering a context-specific network, applies personalized PageRank algorithm and then calculates flux scores of edges. Among edges in the GGN, the union of highly scored edges constructs the context-specific network. Network propagation by personalized PageRank strongly depends on the edge confidence scores and seed node weights. If the reference interactome is an unweighted graph, pyPARAGON sets a default edge score of 1.0 to all edges in the background. Similarly, if seed nodes does not have weights, pyPARAGON assigns them a default value of 1.0 \n",
    "\n",
    "Here, as a guided network, we applied the GGN, which is a trimmed form of reference network. We focus on the related region of reference networks. Reference networks, such as interactomes or regulatomes, are mainly integrated databases that cover a variety of databases regardless of context specificity. Thus, it is critical that GGN selects the associated region of the reference network. However, at that point, the user may use tissue-specific, cell lineage-specific, or other specific networks as a guide, independent of the GGN construction. Another case is that reference networks, such as yeast interactomes, may not be as highly connected as human reference networks. In this case, the user can use the reference network directly without a guide network. The PageRank Flux and Network Inference script details the use of guide networks.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "judicial-passport",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Paragon import NetworkInference "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "wound-logic",
   "metadata": {},
   "outputs": [],
   "source": [
    "pgrf=NetworkInference(network=HIPPIE_nx,guide_network=GGN_nx,edge_attribute=\"Score\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ready-oxford",
   "metadata": {},
   "source": [
    "When invoking the network inference step in pyPARAGON, it is mandatory to introduce a reference network. pyPARAGON assumes the whole reference network functions as a guide network in the absence of a specific guide network introduction. If the user does not provide an edge attribute as a confidence score or another edge score, pyPARAGON assigns a default value of 1 to all edges. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "copyrighted-arbitration",
   "metadata": {},
   "outputs": [],
   "source": [
    "pgrf.load_initial_nodes(list(initial_nodes))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bacterial-receiver",
   "metadata": {},
   "source": [
    "Seed nodes in this script lack weights. Thus, pyPARAGON gives a uniform value of 1 to all seed nodes. The user may input the specified weights of seed nodes ranging from 0 to 1, as explained in the PageRank Flux and Network Inference script.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "further-guinea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "theshold of 0.560000 limits predictions to 2000 edges\n"
     ]
    }
   ],
   "source": [
    "Infered_nx=pgrf.reconstruct_subnetwork( max_edge_count=2000,alpha=0.8,threshold=0.8)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "instructional-sharp",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pgrf.write_created_network(f'../Outputs/AndrogenReceptor_inferred_context_specific_network')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "professional-comedy",
   "metadata": {},
   "source": [
    "# D) Interpreting a context specific network"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "painful-mailing",
   "metadata": {},
   "source": [
    "pyPARAGON partitions a context-specific network into separate communities known as network modules for biological analysis, integrating these communities with biological annotations. This script explores over-representation analysis using gene ontology annotations. The user may expand the use of overrepresentation analysis to additional datasets like KEGG, Reactome, and others. The script of Interpretable_Communities_in_Subnetworks provides a full explanation of the interpretation module, community analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "national-camel",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Paragon import CommunityAnalysis "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "flying-scale",
   "metadata": {},
   "outputs": [],
   "source": [
    "CA=CommunityAnalysis(Infered_nx)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "paperback-vienna",
   "metadata": {},
   "source": [
    "### Community detection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "quality-latino",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genes</th>\n",
       "      <th>Community</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RHOG</td>\n",
       "      <td>Module_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IL6R</td>\n",
       "      <td>Module_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CMYA5</td>\n",
       "      <td>Module_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SMARCA4</td>\n",
       "      <td>Module_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AR</td>\n",
       "      <td>Module_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>MAPK1</td>\n",
       "      <td>Module_32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346</th>\n",
       "      <td>CASP8</td>\n",
       "      <td>Module_32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>347</th>\n",
       "      <td>RB1</td>\n",
       "      <td>Module_32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>348</th>\n",
       "      <td>PARP1</td>\n",
       "      <td>Module_32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>GMEB1</td>\n",
       "      <td>Module_32</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>350 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Genes  Community\n",
       "0       RHOG   Module_1\n",
       "1       IL6R   Module_1\n",
       "2      CMYA5   Module_1\n",
       "3    SMARCA4   Module_1\n",
       "4         AR   Module_1\n",
       "..       ...        ...\n",
       "345    MAPK1  Module_32\n",
       "346    CASP8  Module_32\n",
       "347      RB1  Module_32\n",
       "348    PARP1  Module_32\n",
       "349    GMEB1  Module_32\n",
       "\n",
       "[350 rows x 2 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "module_df, node_df=CA.get_communities_in_DataFrames()\n",
    "node_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "interested-montgomery",
   "metadata": {},
   "outputs": [],
   "source": [
    "node_df.to_csv(f'../Outputs/AndrogenReceptor_inferred_csn_nodes_in_communities.tab',sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "german-indiana",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Community_name</th>\n",
       "      <th>Community</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Module_1</td>\n",
       "      <td>RHOG;IL6R;CMYA5;SMARCA4;AR;ZMIZ1;NR2C1;SRY;TLE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Module_2</td>\n",
       "      <td>RHOB;DNM2;APPL1;PRNP;BSG;VCP;RUVBL2;PRKCZ;TRAF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Module_3</td>\n",
       "      <td>S100A2;PSMC3IP;CDH1;ERBB2;PIK3R2;PIK3CA;CCR5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Module_4</td>\n",
       "      <td>SIRT1;CSNK2A1;HDAC2;SNW1;HIC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Module_5</td>\n",
       "      <td>UBE2E3;CHD3;STK40;SEZ6L2;ETV5;COP1;DET1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Module_6</td>\n",
       "      <td>WWP2;KDM4C;VIRMA;ACTN4;RANGAP1;NELFCD;NXF1;UBA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Module_7</td>\n",
       "      <td>HDAC7;YWHAZ;YWHAB;PPP2CA;YWHAH;FOXO1;RAC3;UBE2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Module_8</td>\n",
       "      <td>GTF2B;NBR1;USP43;NCOA4;MAPK6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Module_9</td>\n",
       "      <td>RHOD;ARRB2;PML;SENP1;ELAVL1;HNRNPK;HDAC4;SUMO3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Module_10</td>\n",
       "      <td>CDKN1A;TRIM63;HGS;PPP1CC;DYRK1A;POLR2A;H2AFB1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Module_11</td>\n",
       "      <td>SATB1;NR2F1;RELA;HIST1H3A;XPO1;RBX1;FOXA1;AHR;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Module_12</td>\n",
       "      <td>VCL;CBLB;TGFB1I1;RNF123;CSK;CRKL;NEDD4L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Module_13</td>\n",
       "      <td>TRIM68;GRAP2;UBE2D3;DISC1;SFPQ;REL;GPS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Module_14</td>\n",
       "      <td>RANBP3;USP7;PIAS1;ZMIZ2;ARSJ;SOX2;PAX9;PIN1;PIAS4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Module_15</td>\n",
       "      <td>TCF4;PHF7;NEUROD1;PRKN;LEF1;RNF14;ZMYND10;CTNNB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Module_16</td>\n",
       "      <td>HOXA11;SIN3A;EFCAB6;TAL1;MNT;BRMS1;HDAC9;EWSR1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Module_17</td>\n",
       "      <td>POLR3F;NRIP1;PNRC1;RARA;NR5A2;SRC;THRB;MAGEA10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Module_18</td>\n",
       "      <td>HNF4A;PGR;RORA;TXNRD2;SP1;RXRA;NCOA2;SP2;NKX3-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Module_19</td>\n",
       "      <td>NR1I3;NCOA3;KLF13;HNF1A;SMAD2;HIF1A;HSD11B2;NF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Module_20</td>\n",
       "      <td>NR1H3;NR1H4;NCOR2;PPARD;CANX;NCOR1;NR1I2;NR0B2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Module_21</td>\n",
       "      <td>GAPDH;KDM1A;SUMO2;MDM2;ESR1;ARNT;HSPB1;HSPA8;B...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Module_22</td>\n",
       "      <td>NPM1;RBM38;BRD4;PARK7;TAT;CFTR;APP;UBE3A;RCHY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Module_23</td>\n",
       "      <td>GSN;FN1;ACTB;GSK3B;BMI1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Module_24</td>\n",
       "      <td>CDK5;TRIM28;RGCC;CDK2;FZR1;FKBP4;CDK1;HSP90AA1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Module_25</td>\n",
       "      <td>NR3C1;DBT;OGT;HNRNPU;PLEKHA4;HEXIM1;BIRC3;HDAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Module_26</td>\n",
       "      <td>CDK14;RBM39;HNRNPL;CIT;DDX5;XRCC5;SUMO1;NTRK1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Module_27</td>\n",
       "      <td>GTF2H5;MKI67;ZNF784;GTF2H4;GTF2H2;SHLD2;GTF2H3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Module_28</td>\n",
       "      <td>E2F1;GTF2H1;TP53BP1;PPARA;ERCC3;GTF2F1;PRMT2;U...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Module_29</td>\n",
       "      <td>BATF3;MYOCD;NCOA6;FOXO4;HOXC11;SMAD3;DACH1;CEB...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Module_30</td>\n",
       "      <td>FOS;PHB;CEBPA;CEBPG;UBR5;SRA1;HSP90AB1;MAPK8;ATF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Module_31</td>\n",
       "      <td>FBXW7;RUNX2;SMURF2;MSX2;TCF15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Module_32</td>\n",
       "      <td>MAPK1;CASP8;RB1;PARP1;GMEB1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Community_name                                          Community\n",
       "0        Module_1  RHOG;IL6R;CMYA5;SMARCA4;AR;ZMIZ1;NR2C1;SRY;TLE...\n",
       "1        Module_2  RHOB;DNM2;APPL1;PRNP;BSG;VCP;RUVBL2;PRKCZ;TRAF...\n",
       "2        Module_3       S100A2;PSMC3IP;CDH1;ERBB2;PIK3R2;PIK3CA;CCR5\n",
       "3        Module_4                      SIRT1;CSNK2A1;HDAC2;SNW1;HIC1\n",
       "4        Module_5            UBE2E3;CHD3;STK40;SEZ6L2;ETV5;COP1;DET1\n",
       "5        Module_6  WWP2;KDM4C;VIRMA;ACTN4;RANGAP1;NELFCD;NXF1;UBA...\n",
       "6        Module_7  HDAC7;YWHAZ;YWHAB;PPP2CA;YWHAH;FOXO1;RAC3;UBE2...\n",
       "7        Module_8                       GTF2B;NBR1;USP43;NCOA4;MAPK6\n",
       "8        Module_9  RHOD;ARRB2;PML;SENP1;ELAVL1;HNRNPK;HDAC4;SUMO3...\n",
       "9       Module_10  CDKN1A;TRIM63;HGS;PPP1CC;DYRK1A;POLR2A;H2AFB1;...\n",
       "10      Module_11  SATB1;NR2F1;RELA;HIST1H3A;XPO1;RBX1;FOXA1;AHR;...\n",
       "11      Module_12            VCL;CBLB;TGFB1I1;RNF123;CSK;CRKL;NEDD4L\n",
       "12      Module_13            TRIM68;GRAP2;UBE2D3;DISC1;SFPQ;REL;GPS2\n",
       "13      Module_14  RANBP3;USP7;PIAS1;ZMIZ2;ARSJ;SOX2;PAX9;PIN1;PIAS4\n",
       "14      Module_15   TCF4;PHF7;NEUROD1;PRKN;LEF1;RNF14;ZMYND10;CTNNB1\n",
       "15      Module_16  HOXA11;SIN3A;EFCAB6;TAL1;MNT;BRMS1;HDAC9;EWSR1...\n",
       "16      Module_17  POLR3F;NRIP1;PNRC1;RARA;NR5A2;SRC;THRB;MAGEA10...\n",
       "17      Module_18  HNF4A;PGR;RORA;TXNRD2;SP1;RXRA;NCOA2;SP2;NKX3-...\n",
       "18      Module_19  NR1I3;NCOA3;KLF13;HNF1A;SMAD2;HIF1A;HSD11B2;NF...\n",
       "19      Module_20  NR1H3;NR1H4;NCOR2;PPARD;CANX;NCOR1;NR1I2;NR0B2...\n",
       "20      Module_21  GAPDH;KDM1A;SUMO2;MDM2;ESR1;ARNT;HSPB1;HSPA8;B...\n",
       "21      Module_22     NPM1;RBM38;BRD4;PARK7;TAT;CFTR;APP;UBE3A;RCHY1\n",
       "22      Module_23                            GSN;FN1;ACTB;GSK3B;BMI1\n",
       "23      Module_24  CDK5;TRIM28;RGCC;CDK2;FZR1;FKBP4;CDK1;HSP90AA1...\n",
       "24      Module_25  NR3C1;DBT;OGT;HNRNPU;PLEKHA4;HEXIM1;BIRC3;HDAC...\n",
       "25      Module_26  CDK14;RBM39;HNRNPL;CIT;DDX5;XRCC5;SUMO1;NTRK1;...\n",
       "26      Module_27  GTF2H5;MKI67;ZNF784;GTF2H4;GTF2H2;SHLD2;GTF2H3...\n",
       "27      Module_28  E2F1;GTF2H1;TP53BP1;PPARA;ERCC3;GTF2F1;PRMT2;U...\n",
       "28      Module_29  BATF3;MYOCD;NCOA6;FOXO4;HOXC11;SMAD3;DACH1;CEB...\n",
       "29      Module_30  FOS;PHB;CEBPA;CEBPG;UBR5;SRA1;HSP90AB1;MAPK8;ATF2\n",
       "30      Module_31                      FBXW7;RUNX2;SMURF2;MSX2;TCF15\n",
       "31      Module_32                        MAPK1;CASP8;RB1;PARP1;GMEB1"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "module_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "demonstrated-chocolate",
   "metadata": {},
   "outputs": [],
   "source": [
    "module_df.to_csv(f'../Outputs/AndrogenReceptor_inferred_csn_communities.tab',sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "express-amino",
   "metadata": {},
   "source": [
    "### Overrepresentation analysis (ORA) with gene ontology annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "colonial-seeking",
   "metadata": {},
   "source": [
    "Reference knowledge for ORA is tabulated in two columns: Specific ID or name of annotations, and components which are represented in networks.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "identical-turkish",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO ID</th>\n",
       "      <th>DB Object Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0002250</td>\n",
       "      <td>IGKV3-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0002250</td>\n",
       "      <td>IGKV1D-42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0002250</td>\n",
       "      <td>IGLV4-69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0002250</td>\n",
       "      <td>IGLV8-61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0002250</td>\n",
       "      <td>IGLV4-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210592</th>\n",
       "      <td>GO:0006958</td>\n",
       "      <td>C1QA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210593</th>\n",
       "      <td>GO:0001682</td>\n",
       "      <td>RPP40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210594</th>\n",
       "      <td>GO:0061640</td>\n",
       "      <td>DCTN3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210595</th>\n",
       "      <td>GO:0045892</td>\n",
       "      <td>NELFCD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210596</th>\n",
       "      <td>GO:0006248</td>\n",
       "      <td>NT5C3A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>210597 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             GO ID DB Object Symbol\n",
       "0       GO:0002250          IGKV3-7\n",
       "1       GO:0002250        IGKV1D-42\n",
       "2       GO:0002250         IGLV4-69\n",
       "3       GO:0002250         IGLV8-61\n",
       "4       GO:0002250         IGLV4-60\n",
       "...            ...              ...\n",
       "210592  GO:0006958             C1QA\n",
       "210593  GO:0001682            RPP40\n",
       "210594  GO:0061640            DCTN3\n",
       "210595  GO:0045892           NELFCD\n",
       "210596  GO:0006248           NT5C3A\n",
       "\n",
       "[210597 rows x 2 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GOA_bio_proc_df=pd.read_csv(f'../Source/Annotations/GOA_proteins_isoforms_prepared.tab',sep=\"\\t\")\n",
    "GOA_bio_proc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "excess-professional",
   "metadata": {},
   "outputs": [],
   "source": [
    "returned_all=CA.hypergeometric_test_for_all_communities(reference_network=HIPPIE_nx,\n",
    "                                           prior_knowledge_df=GOA_bio_proc_df,\n",
    "                                           prior_knowledge_on=\"GO ID\", \n",
    "                                           name_on=\"DB Object Symbol\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "single-gathering",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Community_name</th>\n",
       "      <th>GO ID</th>\n",
       "      <th>p-value</th>\n",
       "      <th>Erichment_Score</th>\n",
       "      <th>Genes in Module</th>\n",
       "      <th>Intersecting Genes</th>\n",
       "      <th>The number of intersecting genes</th>\n",
       "      <th>Process_Gene</th>\n",
       "      <th>The number of components of prior_knowledge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Module_1</td>\n",
       "      <td>GO:0000122</td>\n",
       "      <td>1.522214e-03</td>\n",
       "      <td>3.150793</td>\n",
       "      <td>[RHOG, IL6R, CMYA5, SMARCA4, AR, ZMIZ1, NR2C1,...</td>\n",
       "      <td>[AR, NR2C1, SMARCA4, SRY]</td>\n",
       "      <td>4</td>\n",
       "      <td>[LINC-PINT, E2F8, FEZF1, CNOT1, NUPR2, HELT, N...</td>\n",
       "      <td>968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Module_1</td>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>1.307438e-04</td>\n",
       "      <td>3.636005</td>\n",
       "      <td>[RHOG, IL6R, CMYA5, SMARCA4, AR, ZMIZ1, NR2C1,...</td>\n",
       "      <td>[SRY, NR2C1, AR, SMARCA4, TLE1]</td>\n",
       "      <td>5</td>\n",
       "      <td>[ZNF722, A0A2R8YD15, FOXO3B, NFILZ, A0A7P0TAN4...</td>\n",
       "      <td>1006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Module_1</td>\n",
       "      <td>GO:0006357</td>\n",
       "      <td>1.158394e-02</td>\n",
       "      <td>2.189669</td>\n",
       "      <td>[RHOG, IL6R, CMYA5, SMARCA4, AR, ZMIZ1, NR2C1,...</td>\n",
       "      <td>[SMARCA4, AR, ZMIZ1, NR2C1]</td>\n",
       "      <td>4</td>\n",
       "      <td>[MT-RNR1, EPOP, UNCX, ZNF98, ZNF716, ZNF724, Z...</td>\n",
       "      <td>1697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Module_1</td>\n",
       "      <td>GO:0008284</td>\n",
       "      <td>4.400661e-06</td>\n",
       "      <td>4.737049</td>\n",
       "      <td>[RHOG, IL6R, CMYA5, SMARCA4, AR, ZMIZ1, NR2C1,...</td>\n",
       "      <td>[IL6R, AR, IL6ST, SMARCA4, RHOG]</td>\n",
       "      <td>5</td>\n",
       "      <td>[PRAMEF33, A0A0G2JP20, PRAMEF22, PRAMEF27, NKX...</td>\n",
       "      <td>498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Module_1</td>\n",
       "      <td>GO:0010628</td>\n",
       "      <td>1.885954e-03</td>\n",
       "      <td>3.717017</td>\n",
       "      <td>[RHOG, IL6R, CMYA5, SMARCA4, AR, ZMIZ1, NR2C1,...</td>\n",
       "      <td>[AR, TLE1, SRY]</td>\n",
       "      <td>3</td>\n",
       "      <td>[ODAM, KPNA7, LGALS9, TLR4, BMPR1B, ETV2, PIK3...</td>\n",
       "      <td>460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>Module_30</td>\n",
       "      <td>GO:0140467</td>\n",
       "      <td>1.775290e-07</td>\n",
       "      <td>8.434220</td>\n",
       "      <td>[FOS, PHB, CEBPA, CEBPG, UBR5, SRA1, HSP90AB1,...</td>\n",
       "      <td>[FOS, CEBPA, CEBPG]</td>\n",
       "      <td>3</td>\n",
       "      <td>[CREB3, MAF, FOS, JUN, FOSL1, JUNB, CEBPB, ATF...</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>Module_31</td>\n",
       "      <td>GO:0006357</td>\n",
       "      <td>5.805930e-03</td>\n",
       "      <td>3.318940</td>\n",
       "      <td>[FBXW7, RUNX2, SMURF2, MSX2, TCF15]</td>\n",
       "      <td>[RUNX2, TCF15, MSX2]</td>\n",
       "      <td>3</td>\n",
       "      <td>[MT-RNR1, EPOP, UNCX, ZNF98, ZNF716, ZNF724, Z...</td>\n",
       "      <td>1697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>Module_31</td>\n",
       "      <td>GO:0045892</td>\n",
       "      <td>2.618700e-04</td>\n",
       "      <td>4.987517</td>\n",
       "      <td>[FBXW7, RUNX2, SMURF2, MSX2, TCF15]</td>\n",
       "      <td>[MSX2, RUNX2, SMURF2]</td>\n",
       "      <td>3</td>\n",
       "      <td>[PRAMEF33, A0A0G2JP20, TLE7, PRAMEF22, PRAMEF2...</td>\n",
       "      <td>587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>Module_32</td>\n",
       "      <td>GO:0006915</td>\n",
       "      <td>2.658497e-04</td>\n",
       "      <td>4.979789</td>\n",
       "      <td>[MAPK1, CASP8, RB1, PARP1, GMEB1]</td>\n",
       "      <td>[PARP1, MAPK1, CASP8]</td>\n",
       "      <td>3</td>\n",
       "      <td>[SYCE3, USP17L3, USP17L5, LGALS16, KLLN, PANO1...</td>\n",
       "      <td>590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>Module_32</td>\n",
       "      <td>GO:0045944</td>\n",
       "      <td>2.136107e-03</td>\n",
       "      <td>3.880821</td>\n",
       "      <td>[MAPK1, CASP8, RB1, PARP1, GMEB1]</td>\n",
       "      <td>[RB1, PARP1, GMEB1]</td>\n",
       "      <td>3</td>\n",
       "      <td>[E2F8, ZBTB7C, NKX2-6, ZNF840P, HELT, MSGN1, N...</td>\n",
       "      <td>1200</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>253 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Community_name       GO ID       p-value  Erichment_Score  \\\n",
       "0         Module_1  GO:0000122  1.522214e-03         3.150793   \n",
       "1         Module_1  GO:0006355  1.307438e-04         3.636005   \n",
       "2         Module_1  GO:0006357  1.158394e-02         2.189669   \n",
       "3         Module_1  GO:0008284  4.400661e-06         4.737049   \n",
       "4         Module_1  GO:0010628  1.885954e-03         3.717017   \n",
       "..             ...         ...           ...              ...   \n",
       "248      Module_30  GO:0140467  1.775290e-07         8.434220   \n",
       "249      Module_31  GO:0006357  5.805930e-03         3.318940   \n",
       "250      Module_31  GO:0045892  2.618700e-04         4.987517   \n",
       "251      Module_32  GO:0006915  2.658497e-04         4.979789   \n",
       "252      Module_32  GO:0045944  2.136107e-03         3.880821   \n",
       "\n",
       "                                       Genes in Module  \\\n",
       "0    [RHOG, IL6R, CMYA5, SMARCA4, AR, ZMIZ1, NR2C1,...   \n",
       "1    [RHOG, IL6R, CMYA5, SMARCA4, AR, ZMIZ1, NR2C1,...   \n",
       "2    [RHOG, IL6R, CMYA5, SMARCA4, AR, ZMIZ1, NR2C1,...   \n",
       "3    [RHOG, IL6R, CMYA5, SMARCA4, AR, ZMIZ1, NR2C1,...   \n",
       "4    [RHOG, IL6R, CMYA5, SMARCA4, AR, ZMIZ1, NR2C1,...   \n",
       "..                                                 ...   \n",
       "248  [FOS, PHB, CEBPA, CEBPG, UBR5, SRA1, HSP90AB1,...   \n",
       "249                [FBXW7, RUNX2, SMURF2, MSX2, TCF15]   \n",
       "250                [FBXW7, RUNX2, SMURF2, MSX2, TCF15]   \n",
       "251                  [MAPK1, CASP8, RB1, PARP1, GMEB1]   \n",
       "252                  [MAPK1, CASP8, RB1, PARP1, GMEB1]   \n",
       "\n",
       "                   Intersecting Genes The number of intersecting genes  \\\n",
       "0           [AR, NR2C1, SMARCA4, SRY]                                4   \n",
       "1     [SRY, NR2C1, AR, SMARCA4, TLE1]                                5   \n",
       "2         [SMARCA4, AR, ZMIZ1, NR2C1]                                4   \n",
       "3    [IL6R, AR, IL6ST, SMARCA4, RHOG]                                5   \n",
       "4                     [AR, TLE1, SRY]                                3   \n",
       "..                                ...                              ...   \n",
       "248               [FOS, CEBPA, CEBPG]                                3   \n",
       "249              [RUNX2, TCF15, MSX2]                                3   \n",
       "250             [MSX2, RUNX2, SMURF2]                                3   \n",
       "251             [PARP1, MAPK1, CASP8]                                3   \n",
       "252               [RB1, PARP1, GMEB1]                                3   \n",
       "\n",
       "                                          Process_Gene  \\\n",
       "0    [LINC-PINT, E2F8, FEZF1, CNOT1, NUPR2, HELT, N...   \n",
       "1    [ZNF722, A0A2R8YD15, FOXO3B, NFILZ, A0A7P0TAN4...   \n",
       "2    [MT-RNR1, EPOP, UNCX, ZNF98, ZNF716, ZNF724, Z...   \n",
       "3    [PRAMEF33, A0A0G2JP20, PRAMEF22, PRAMEF27, NKX...   \n",
       "4    [ODAM, KPNA7, LGALS9, TLR4, BMPR1B, ETV2, PIK3...   \n",
       "..                                                 ...   \n",
       "248  [CREB3, MAF, FOS, JUN, FOSL1, JUNB, CEBPB, ATF...   \n",
       "249  [MT-RNR1, EPOP, UNCX, ZNF98, ZNF716, ZNF724, Z...   \n",
       "250  [PRAMEF33, A0A0G2JP20, TLE7, PRAMEF22, PRAMEF2...   \n",
       "251  [SYCE3, USP17L3, USP17L5, LGALS16, KLLN, PANO1...   \n",
       "252  [E2F8, ZBTB7C, NKX2-6, ZNF840P, HELT, MSGN1, N...   \n",
       "\n",
       "    The number of components of prior_knowledge  \n",
       "0                                           968  \n",
       "1                                          1006  \n",
       "2                                          1697  \n",
       "3                                           498  \n",
       "4                                           460  \n",
       "..                                          ...  \n",
       "248                                          26  \n",
       "249                                        1697  \n",
       "250                                         587  \n",
       "251                                         590  \n",
       "252                                        1200  \n",
       "\n",
       "[253 rows x 9 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "returned_all "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "studied-gentleman",
   "metadata": {},
   "outputs": [],
   "source": [
    "returned_all.to_csv(f'../Outputs/AndrogenReceptor_inferred_csn_overrepresentation_results.tab',sep=\"\\t\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
