{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "heated-strip",
   "metadata": {},
   "source": [
    "# A. Input data-sets\n",
    "\n",
    "pyPARAGON requires two inputs: \n",
    "\n",
    "    i) A reference network which is composed of already known interactions, \n",
    "    ii) Seed nodes which can come from in omics datasets, or be frequent mutations in disases. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "saving-teach",
   "metadata": {},
   "source": [
    "### Preparing a reference network\n",
    "\n",
    "A reference network is introduced to pyPARAGON as a networkx object. Reference networks can include nan-nan interactions, causing troubles in calculations. Thus, we exclude nan-nan interactions in reference networks. Additionally, pyPARAGON does not consider self-interactions in graphlets. Therefore, we also eliminate self interactions to get rid of computational problems. Except for nan-nan and self-interactions, we may not use any filtration or additional preprocessing steps. pyPARAGON is able to eliminate the interactions with low confidence scores thanks to the flux calculation. Thus, we suggest not applying a confidence threshold value for the filtration. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "mineral-separate",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd, networkx as nx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "handy-retailer",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_1</th>\n",
       "      <th>Gene_2</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ALDH1A1</td>\n",
       "      <td>ALDH1A1</td>\n",
       "      <td>0.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ITGA7</td>\n",
       "      <td>CHRNA1</td>\n",
       "      <td>0.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PPP1R9A</td>\n",
       "      <td>ACTG1</td>\n",
       "      <td>0.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRGN</td>\n",
       "      <td>CD44</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GRB7</td>\n",
       "      <td>ERBB2</td>\n",
       "      <td>0.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783177</th>\n",
       "      <td>STIL</td>\n",
       "      <td>CDK5RAP2</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783178</th>\n",
       "      <td>SIRT7</td>\n",
       "      <td>SNORA10</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783179</th>\n",
       "      <td>BMP2</td>\n",
       "      <td>SNX18</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783180</th>\n",
       "      <td>C1QBP</td>\n",
       "      <td>MRPL50</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783181</th>\n",
       "      <td>CCNF</td>\n",
       "      <td>LMNB2</td>\n",
       "      <td>0.63</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>783182 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene_1    Gene_2  Score\n",
       "0       ALDH1A1   ALDH1A1   0.76\n",
       "1         ITGA7    CHRNA1   0.73\n",
       "2       PPP1R9A     ACTG1   0.65\n",
       "3          SRGN      CD44   0.63\n",
       "4          GRB7     ERBB2   0.90\n",
       "...         ...       ...    ...\n",
       "783177     STIL  CDK5RAP2   0.63\n",
       "783178    SIRT7   SNORA10   0.63\n",
       "783179     BMP2     SNX18   0.63\n",
       "783180    C1QBP    MRPL50   0.63\n",
       "783181     CCNF     LMNB2   0.63\n",
       "\n",
       "[783182 rows x 3 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HIPPIE_df=pd.read_csv(\"../Source/Interactomes/HIPPIE_v2_3.tab\", keep_default_na=False, sep=\"\\t\") # excluding rows with nan value\n",
    "HIPPIE_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "desperate-things",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "tested-april",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of edges in HIPPIE network 774448\n",
      "The number of nodes in HIPPIE network 19437\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Networkx object and removing self interactions\n",
    "HIPPIE_nx=nx.from_pandas_edgelist(HIPPIE_df,\"Gene_1\",\"Gene_2\",edge_attr=\"Score\")\n",
    "self_interactions=list(nx.selfloop_edges(HIPPIE_nx))\n",
    "\n",
    "\n",
    "for u, v in self_interactions:\n",
    "    if isinstance(u, float):\n",
    "        print(u)\n",
    "    else:\n",
    "        HIPPIE_nx.remove_edge(u,v)\n",
    "    \n",
    "HIPPIE_df=nx.to_pandas_edgelist(HIPPIE_nx)\n",
    "print(\"The number of edges in HIPPIE network\",HIPPIE_nx.number_of_edges())\n",
    "print(\"The number of nodes in HIPPIE network\",HIPPIE_nx.number_of_nodes())\n",
    "\n",
    "\n",
    "edges=tuple(zip(HIPPIE_df.source,HIPPIE_df.target))\n",
    "for edge in edges:\n",
    "    if edge not in HIPPIE_nx.edges:\n",
    "        print(\"check the edge\",edge)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "expired-tours",
   "metadata": {},
   "source": [
    "###  Preparing seed nodes\n",
    "\n",
    "Seed nodes are known or detected context-specific knowledge. They can come from omics datasets, multi-omics datasets, or well-known, frequent mutations in diseases. Score your seeds as 1 if you do not generate initial node weights from raw data.\n",
    "\n",
    "During multi-omics data integration, the user can follow two strategies with pyPARAGON. In the first strategy,  your hits in omics datasets can be transformed into nodes represented in the reference network. Identifying possible transcription factors which target differential hits in transcriptomics or bind to the genome can be useful to carry transcriptomics and genomics knowledge into the interactomes.  With the second use option of pyPARAGON in multi-omic datasets, you may extend the reference network by merging various types of networks, such as regulator networks, interactomes, metabolic networks, etc. In this case, you can directly use multi-omics hits, which are represented in the reference network. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "chemical-fifty",
   "metadata": {},
   "outputs": [],
   "source": [
    "initial_nodes_df=pd.read_csv(f'../Source/Netpath/Sampling_0_5/AndrogenReceptor_05A/AndrogenReceptor_05A_var_0.nodes',sep=\"\\t\")\n",
    "\n",
    "initial_nodes=initial_nodes_df.name.to_list()\n",
    "initial_weights=[1 for _ in initial_nodes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "political-tension",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "celtic-being",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "featured-offense",
   "metadata": {},
   "source": [
    "## B) Constructing Graphlet Guided Network \n",
    "\n",
    "During the construction of a graphlet-guided network (GGN), pyPARAGON does not consider the confidence scores of edges in parsing the graphlets. The advanced usage of pyPARAGON in the construction of GGN is detailed in the Graphlet_Guided_Network_Construction, covering reference network permutations, graphlet frequencies, and graphlet motif selections. In this script, we recruit four-node graphlets, such as Graphlet5, Graphlet6, Graphlet7, and Graphlet8. During graphlet generation, we derived four-node graphlets from three-node graphlets. Also you can directly select four-node graphlets in pyPARAGON to overcome the computational cost. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "northern-threshold",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Paragon import GraphletGuidance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "raising-minutes",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "tight-coffee",
   "metadata": {},
   "outputs": [],
   "source": [
    "Graphlet_list=['Graphlets5', 'Graphlets6', 'Graphlets7', 'Graphlets8']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "involved-briefing",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "GRF_lite=GraphletGuidance(HIPPIE_nx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "missing-tuning",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "82 of 82 input nodes have been found in the given network\n",
      "\n"
     ]
    }
   ],
   "source": [
    "GGN_nx=GRF_lite.construct_GGN(initial_nodes,Graphlets=Graphlet_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "designing-world",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4137"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GGN_nx.number_of_nodes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "numeric-ethnic",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CEBPA</td>\n",
       "      <td>SMARCA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CEBPA</td>\n",
       "      <td>ESR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CEBPA</td>\n",
       "      <td>SMAD3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CEBPA</td>\n",
       "      <td>CDK4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CEBPA</td>\n",
       "      <td>ATF2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20805</th>\n",
       "      <td>SPDEF</td>\n",
       "      <td>SLC30A9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20806</th>\n",
       "      <td>SPDEF</td>\n",
       "      <td>CEP57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20807</th>\n",
       "      <td>SPDEF</td>\n",
       "      <td>KLHL9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20808</th>\n",
       "      <td>SPDEF</td>\n",
       "      <td>SLC4A2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20809</th>\n",
       "      <td>SPDEF</td>\n",
       "      <td>TTC23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20810 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      source   target\n",
       "0      CEBPA  SMARCA2\n",
       "1      CEBPA     ESR1\n",
       "2      CEBPA    SMAD3\n",
       "3      CEBPA     CDK4\n",
       "4      CEBPA     ATF2\n",
       "...      ...      ...\n",
       "20805  SPDEF  SLC30A9\n",
       "20806  SPDEF    CEP57\n",
       "20807  SPDEF    KLHL9\n",
       "20808  SPDEF   SLC4A2\n",
       "20809  SPDEF    TTC23\n",
       "\n",
       "[20810 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GGN_df=nx.to_pandas_edgelist(GGN_nx)\n",
    "GGN_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "superb-exclusion",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GRF_lite.write_guided_graphlet_network(f'../Outputs/AndrogenReceptor_GGN')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "waiting-opening",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "tired-albuquerque",
   "metadata": {},
   "source": [
    "## C) PageRank flux and context-specific network inference"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "finnish-completion",
   "metadata": {},
   "source": [
    "pyPARAGON, before infering a context-specific network, applies personalized PageRank algorithm and then calculates flux scores of edges. Among edges in the GGN, the union of highly scored edges constructs the context-specific network. In the application of personalized PageRank, the confidence scores and the weights of seed nodes affects the node propagation. When confidence scores or weights of edges in reference networks are not employed, pyPARAGON sets a default edge score of 1 to all edges. Similarly, if seed nodes are not weighted, pyPARAGON assigned default value, 1, to all seed nodes. \n",
    "\n",
    "Here, as a guided network, we applied the GGN, which is a trimmed form of reference network. We focus on the related region of reference networks. Reference networks, such as interactomes or regulatomes, are mainly integrated databases that cover a variety of databases without regard for context specificity. Thus, GGN selects the associated region of the reference network. However, at that point, the user may use tissue-specific, cell lineage-specific, or other guided networks as a guide, regardless of the GGN construction. Another case is that reference networks, such as yeast interactomes, may not be as highly connected as human reference networks. In this case, the user can use the reference network directly without a guide network. The PageRank Flux and Network Inference script details the use of guide networks.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "judicial-passport",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Paragon import NetworkInference "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "wound-logic",
   "metadata": {},
   "outputs": [],
   "source": [
    "pgrf=NetworkInference(network=HIPPIE_nx,guide_network=GGN_nx,edge_attribute=\"Score\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ready-oxford",
   "metadata": {},
   "source": [
    "When invoking the network inference step in pyPARAGON, it is mandatory to introduce a reference network. pyPARAGON assumes the whole reference network functions as a guide network in the absence of a specific guide network introduction. If the user does not provide an edge attribute as a confidence score or another edge score, pyPARAGON assigns a default value of 1 to all edges. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "copyrighted-arbitration",
   "metadata": {},
   "outputs": [],
   "source": [
    "pgrf.load_initial_nodes(list(initial_nodes))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bacterial-receiver",
   "metadata": {},
   "source": [
    "Seed nodes in this script lack weights. Thus, pyPARAGON gives a uniform value of 1 to all seed nodes. The user may input the specified weights of seed nodes ranging from 0 to 1, as explained in the PageRank Flux and Network Inference script.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "further-guinea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "theshold of 0.560000 limits predictions to 2000 edges\n"
     ]
    }
   ],
   "source": [
    "Infered_nx=pgrf.reconstruct_subnetwork( max_edge_count=2000,alpha=0.8,threshold=0.8)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "instructional-sharp",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pgrf.write_created_network(f'../Outputs/AndrogenReceptor_inferred_context_specific_network')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "available-jenny",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dressed-protection",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "professional-comedy",
   "metadata": {},
   "source": [
    "# D) Interpreting a context specific network"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "painful-mailing",
   "metadata": {},
   "source": [
    "pyPARAGON partitions a context-specific network into separate communities known as network modules for biological analysis, integrating these communities with biological annotations. This script explores over-representation analysis using gene ontology annotations. The user may expand the use of overrepresentation analysis to additional datasets like KEGG, Reactome, and others. The script of Interpretable_Communities_in_Subnetworks provides a full explanation of the interpretation module, community analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "national-camel",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Paragon import CommunityAnalysis "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "flying-scale",
   "metadata": {},
   "outputs": [],
   "source": [
    "CA=CommunityAnalysis(Infered_nx)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "paperback-vienna",
   "metadata": {},
   "source": [
    "### Community detection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "quality-latino",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genes</th>\n",
       "      <th>Community</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RNF14</td>\n",
       "      <td>Module_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UBE2E3</td>\n",
       "      <td>Module_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PHF7</td>\n",
       "      <td>Module_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DET1</td>\n",
       "      <td>Module_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LEF1</td>\n",
       "      <td>Module_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>UBD</td>\n",
       "      <td>Module_33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>RAC3</td>\n",
       "      <td>Module_33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>CDC42</td>\n",
       "      <td>Module_33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>ARRB2</td>\n",
       "      <td>Module_33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>RHOD</td>\n",
       "      <td>Module_33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>357 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Genes  Community\n",
       "0     RNF14   Module_1\n",
       "1    UBE2E3   Module_1\n",
       "2      PHF7   Module_1\n",
       "3      DET1   Module_1\n",
       "4      LEF1   Module_1\n",
       "..      ...        ...\n",
       "352     UBD  Module_33\n",
       "353    RAC3  Module_33\n",
       "354   CDC42  Module_33\n",
       "355   ARRB2  Module_33\n",
       "356    RHOD  Module_33\n",
       "\n",
       "[357 rows x 2 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "module_df, node_df=CA.get_communities_in_DataFrames()\n",
    "node_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "opening-eight",
   "metadata": {},
   "outputs": [],
   "source": [
    "node_df.to_csv(f'../Outputs/AndrogenReceptor_inferred_csn_nodes_in_communities.tab',sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "german-indiana",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Community_name</th>\n",
       "      <th>Community</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Module_1</td>\n",
       "      <td>RNF14;UBE2E3;PHF7;DET1;LEF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Module_2</td>\n",
       "      <td>DISC1;SFPQ;REL;UBE2D3;GPS2;TRIM68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Module_3</td>\n",
       "      <td>ZMYND10;CALM1;UBE2I;HIC1;TCF4;OGT;NEUROD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Module_4</td>\n",
       "      <td>GRAP2;STK40;NXF1;ETV5;COP1;CHD3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Module_5</td>\n",
       "      <td>PRDM11;ERCC3;GTF2F1;GTF2H1;E2F1;PPARA;FGL1;PIK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Module_6</td>\n",
       "      <td>TLE1;GTF2H2C;IL6R;CMYA5;AR;NR2C1;IL6ST;ERBB2;P...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Module_7</td>\n",
       "      <td>GTF2H2;CTD;SHLD2;GTF2H3;CCNH;MKI67;GTF2E1;ERCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Module_8</td>\n",
       "      <td>GRB2;NTRK1;CIT;MAPT;GSK3B;ACTB;HSP90AB1;CDK5;F...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Module_9</td>\n",
       "      <td>RAN;SUMO1;HSPA1A;UBE3A;PXN;EP300;GAPDH;HSPA8;C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Module_10</td>\n",
       "      <td>ESR1;BRCA1;NR3C1;ARNT;AKT1;TP53;HSP90AA1;HSPB1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Module_11</td>\n",
       "      <td>EFTUD2;DLD;RHOG;PARK7;NFKB1;PLEKHA4;HDAC6;DBT;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Module_12</td>\n",
       "      <td>MDM2;XRCC5;HDAC2;SUMO2;XRCC6;CDK9;EWSR1;TRIM25...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Module_13</td>\n",
       "      <td>HOXA11;CTNNB1;YWHAH;PPP2CA;YWHAB;YWHAZ;YWHAQ;S...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Module_14</td>\n",
       "      <td>CDK4;PCNA;TSC2;DYRK1A;CDKN1A;CDKN1B;CDK11B;CDK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Module_15</td>\n",
       "      <td>MAPK8;BATF3;UBR5;RB1;CEBPA;SRA1;JUN;ATF2;ETS1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Module_16</td>\n",
       "      <td>CANX;VIM;KDM4C;APEX1;VIRMA;UBA1;SMAD4;RANGAP1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Module_17</td>\n",
       "      <td>SRY;FOXO4;CEBPB;TCF15;HOXC11;RUNX2;SMURF2;SMAD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Module_18</td>\n",
       "      <td>RBX1;RNF123;AHR;NR2F1;XPO1;FOXA1;SEZ6L2;FHL2;C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Module_19</td>\n",
       "      <td>PPP1CA;VCL;NEDD4L;PIAS3;CSK;YWHAG;SUMO3;TGFB1I...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Module_20</td>\n",
       "      <td>SPDEF;NCOA4;UXT;PELP1;POLR2A;GTF2B;ECT2;MECOM;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Module_21</td>\n",
       "      <td>SMARCA2;MEOX2;BRMS1;KLF1;SMARCE1;RARB;MEF2C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Module_22</td>\n",
       "      <td>EFCAB6;DACH1;NCOR1;CBFA2T3;NFE4;PML;PPARD;JDP2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Module_23</td>\n",
       "      <td>SIN3A;EZH2;HDAC9;PHB;KLF13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Module_24</td>\n",
       "      <td>BAG3;VDR;NCOR2;H2AFB1;BAG1;HDAC3;ABL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Module_25</td>\n",
       "      <td>THRA;RXRA;HIF1A;THRB;SMAD2;NCOA2;PPARG;NCOA1;N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Module_26</td>\n",
       "      <td>NRIP1;RXRB;NR1H2;NR1I3;RORB;NR1I2;NR3C2;NR1H3;SRC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Module_27</td>\n",
       "      <td>PGR;MAGEA10;HSD11B2;HNF1A;KAT2B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Module_28</td>\n",
       "      <td>RORA;SMARCA4;SP1;MNT;RNF4;SP2;ONECUT1;ESRRA;NK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Module_29</td>\n",
       "      <td>ESRRG;NR0B2;RARA;CIDEC;HNF4A;SMARCB1;HNF4G;PNR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Module_30</td>\n",
       "      <td>PSMC3IP;RBAK;CASP8;SIRT1;ZMIZ1;SNW1;S100A2;EIF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Module_31</td>\n",
       "      <td>DNM2;VCP;HDAC5;RHOB;PRKCZ;BSG;APPL1;PRNP;RUVBL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Module_32</td>\n",
       "      <td>SOX2;PAX9;PIN1;RANBP3;ZMIZ2;PIAS1;PIAS4;ARSJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Module_33</td>\n",
       "      <td>HDAC4;ELAVL1;SENP1;RAC2;UBD;RAC3;CDC42;ARRB2;RHOD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Community_name                                          Community\n",
       "0        Module_1                        RNF14;UBE2E3;PHF7;DET1;LEF1\n",
       "1        Module_2                  DISC1;SFPQ;REL;UBE2D3;GPS2;TRIM68\n",
       "2        Module_3          ZMYND10;CALM1;UBE2I;HIC1;TCF4;OGT;NEUROD1\n",
       "3        Module_4                    GRAP2;STK40;NXF1;ETV5;COP1;CHD3\n",
       "4        Module_5  PRDM11;ERCC3;GTF2F1;GTF2H1;E2F1;PPARA;FGL1;PIK...\n",
       "5        Module_6  TLE1;GTF2H2C;IL6R;CMYA5;AR;NR2C1;IL6ST;ERBB2;P...\n",
       "6        Module_7  GTF2H2;CTD;SHLD2;GTF2H3;CCNH;MKI67;GTF2E1;ERCC...\n",
       "7        Module_8  GRB2;NTRK1;CIT;MAPT;GSK3B;ACTB;HSP90AB1;CDK5;F...\n",
       "8        Module_9  RAN;SUMO1;HSPA1A;UBE3A;PXN;EP300;GAPDH;HSPA8;C...\n",
       "9       Module_10  ESR1;BRCA1;NR3C1;ARNT;AKT1;TP53;HSP90AA1;HSPB1...\n",
       "10      Module_11  EFTUD2;DLD;RHOG;PARK7;NFKB1;PLEKHA4;HDAC6;DBT;...\n",
       "11      Module_12  MDM2;XRCC5;HDAC2;SUMO2;XRCC6;CDK9;EWSR1;TRIM25...\n",
       "12      Module_13  HOXA11;CTNNB1;YWHAH;PPP2CA;YWHAB;YWHAZ;YWHAQ;S...\n",
       "13      Module_14  CDK4;PCNA;TSC2;DYRK1A;CDKN1A;CDKN1B;CDK11B;CDK...\n",
       "14      Module_15  MAPK8;BATF3;UBR5;RB1;CEBPA;SRA1;JUN;ATF2;ETS1;...\n",
       "15      Module_16  CANX;VIM;KDM4C;APEX1;VIRMA;UBA1;SMAD4;RANGAP1;...\n",
       "16      Module_17  SRY;FOXO4;CEBPB;TCF15;HOXC11;RUNX2;SMURF2;SMAD...\n",
       "17      Module_18  RBX1;RNF123;AHR;NR2F1;XPO1;FOXA1;SEZ6L2;FHL2;C...\n",
       "18      Module_19  PPP1CA;VCL;NEDD4L;PIAS3;CSK;YWHAG;SUMO3;TGFB1I...\n",
       "19      Module_20  SPDEF;NCOA4;UXT;PELP1;POLR2A;GTF2B;ECT2;MECOM;...\n",
       "20      Module_21        SMARCA2;MEOX2;BRMS1;KLF1;SMARCE1;RARB;MEF2C\n",
       "21      Module_22  EFCAB6;DACH1;NCOR1;CBFA2T3;NFE4;PML;PPARD;JDP2...\n",
       "22      Module_23                         SIN3A;EZH2;HDAC9;PHB;KLF13\n",
       "23      Module_24              BAG3;VDR;NCOR2;H2AFB1;BAG1;HDAC3;ABL1\n",
       "24      Module_25  THRA;RXRA;HIF1A;THRB;SMAD2;NCOA2;PPARG;NCOA1;N...\n",
       "25      Module_26  NRIP1;RXRB;NR1H2;NR1I3;RORB;NR1I2;NR3C2;NR1H3;SRC\n",
       "26      Module_27                    PGR;MAGEA10;HSD11B2;HNF1A;KAT2B\n",
       "27      Module_28  RORA;SMARCA4;SP1;MNT;RNF4;SP2;ONECUT1;ESRRA;NK...\n",
       "28      Module_29  ESRRG;NR0B2;RARA;CIDEC;HNF4A;SMARCB1;HNF4G;PNR...\n",
       "29      Module_30  PSMC3IP;RBAK;CASP8;SIRT1;ZMIZ1;SNW1;S100A2;EIF...\n",
       "30      Module_31  DNM2;VCP;HDAC5;RHOB;PRKCZ;BSG;APPL1;PRNP;RUVBL...\n",
       "31      Module_32       SOX2;PAX9;PIN1;RANBP3;ZMIZ2;PIAS1;PIAS4;ARSJ\n",
       "32      Module_33  HDAC4;ELAVL1;SENP1;RAC2;UBD;RAC3;CDC42;ARRB2;RHOD"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "module_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "jewish-public",
   "metadata": {},
   "outputs": [],
   "source": [
    "module_df.to_csv(f'../Outputs/AndrogenReceptor_inferred_csn_communities.tab',sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "express-amino",
   "metadata": {},
   "source": [
    "### Overrepresentation analysis (ORA) with gene ontology annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "colonial-seeking",
   "metadata": {},
   "source": [
    "Reference knowledge for ORA is tabulated in two columns: Specific ID or name of annotations, and components which are represented in networks.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "identical-turkish",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO ID</th>\n",
       "      <th>DB Object Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0002250</td>\n",
       "      <td>IGKV3-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0002250</td>\n",
       "      <td>IGKV1D-42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0002250</td>\n",
       "      <td>IGLV4-69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0002250</td>\n",
       "      <td>IGLV8-61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0002250</td>\n",
       "      <td>IGLV4-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210592</th>\n",
       "      <td>GO:0006958</td>\n",
       "      <td>C1QA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210593</th>\n",
       "      <td>GO:0001682</td>\n",
       "      <td>RPP40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210594</th>\n",
       "      <td>GO:0061640</td>\n",
       "      <td>DCTN3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210595</th>\n",
       "      <td>GO:0045892</td>\n",
       "      <td>NELFCD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210596</th>\n",
       "      <td>GO:0006248</td>\n",
       "      <td>NT5C3A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>210597 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             GO ID DB Object Symbol\n",
       "0       GO:0002250          IGKV3-7\n",
       "1       GO:0002250        IGKV1D-42\n",
       "2       GO:0002250         IGLV4-69\n",
       "3       GO:0002250         IGLV8-61\n",
       "4       GO:0002250         IGLV4-60\n",
       "...            ...              ...\n",
       "210592  GO:0006958             C1QA\n",
       "210593  GO:0001682            RPP40\n",
       "210594  GO:0061640            DCTN3\n",
       "210595  GO:0045892           NELFCD\n",
       "210596  GO:0006248           NT5C3A\n",
       "\n",
       "[210597 rows x 2 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GOA_bio_proc_df=pd.read_csv(f'../Source/Annotations/GOA_proteins_isoforms_prepared.tab',sep=\"\\t\")\n",
    "GOA_bio_proc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "excess-professional",
   "metadata": {},
   "outputs": [],
   "source": [
    "returned_all=CA.hypergeometric_test_for_all_communities(reference_network=HIPPIE_nx,\n",
    "                                           prior_knowledge_df=GOA_bio_proc_df,\n",
    "                                           prior_knowledge_on=\"GO ID\", \n",
    "                                           name_on=\"DB Object Symbol\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "solar-sullivan",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "single-gathering",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Community_name</th>\n",
       "      <th>GO ID</th>\n",
       "      <th>p-value</th>\n",
       "      <th>Erichment_Score</th>\n",
       "      <th>Genes in Module</th>\n",
       "      <th>Intersecting Genes</th>\n",
       "      <th>The number of intersecting genes</th>\n",
       "      <th>Process_Gene</th>\n",
       "      <th>The number of components of prior_knowledge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Module_2</td>\n",
       "      <td>GO:0000122</td>\n",
       "      <td>2.198714e-03</td>\n",
       "      <td>3.785060</td>\n",
       "      <td>[DISC1, SFPQ, REL, UBE2D3, GPS2, TRIM68]</td>\n",
       "      <td>[SFPQ, UBE2D3, GPS2]</td>\n",
       "      <td>3</td>\n",
       "      <td>[LINC-PINT, E2F8, FEZF1, CNOT1, NUPR2, HELT, N...</td>\n",
       "      <td>968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Module_2</td>\n",
       "      <td>GO:0045087</td>\n",
       "      <td>5.068671e-04</td>\n",
       "      <td>4.566379</td>\n",
       "      <td>[DISC1, SFPQ, REL, UBE2D3, GPS2, TRIM68]</td>\n",
       "      <td>[SFPQ, TRIM68, REL]</td>\n",
       "      <td>3</td>\n",
       "      <td>[TRGV2, DEFB131B, TRGV5, TRGV8, TRGV4, TRDV1, ...</td>\n",
       "      <td>585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Module_2</td>\n",
       "      <td>GO:0045944</td>\n",
       "      <td>4.076125e-03</td>\n",
       "      <td>3.441766</td>\n",
       "      <td>[DISC1, SFPQ, REL, UBE2D3, GPS2, TRIM68]</td>\n",
       "      <td>[SFPQ, REL, GPS2]</td>\n",
       "      <td>3</td>\n",
       "      <td>[E2F8, ZBTB7C, NKX2-6, ZNF840P, HELT, MSGN1, N...</td>\n",
       "      <td>1200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Module_3</td>\n",
       "      <td>GO:0000122</td>\n",
       "      <td>3.705554e-03</td>\n",
       "      <td>3.443890</td>\n",
       "      <td>[ZMYND10, CALM1, UBE2I, HIC1, TCF4, OGT, NEUROD1]</td>\n",
       "      <td>[OGT, UBE2I, HIC1]</td>\n",
       "      <td>3</td>\n",
       "      <td>[LINC-PINT, E2F8, FEZF1, CNOT1, NUPR2, HELT, N...</td>\n",
       "      <td>968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Module_3</td>\n",
       "      <td>GO:0006357</td>\n",
       "      <td>1.633251e-03</td>\n",
       "      <td>3.305400</td>\n",
       "      <td>[ZMYND10, CALM1, UBE2I, HIC1, TCF4, OGT, NEUROD1]</td>\n",
       "      <td>[OGT, NEUROD1, TCF4, HIC1]</td>\n",
       "      <td>4</td>\n",
       "      <td>[MT-RNR1, EPOP, UNCX, ZNF98, ZNF716, ZNF724, Z...</td>\n",
       "      <td>1697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>Module_33</td>\n",
       "      <td>GO:0008360</td>\n",
       "      <td>3.730783e-05</td>\n",
       "      <td>5.776614</td>\n",
       "      <td>[HDAC4, ELAVL1, SENP1, RAC2, UBD, RAC3, CDC42,...</td>\n",
       "      <td>[RAC2, RAC3, RHOD]</td>\n",
       "      <td>3</td>\n",
       "      <td>[PALM3, PALM2AKAP2, CCL24, LST1, CDC42EP2, SLC...</td>\n",
       "      <td>151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>Module_33</td>\n",
       "      <td>GO:0010592</td>\n",
       "      <td>2.493232e-07</td>\n",
       "      <td>8.263737</td>\n",
       "      <td>[HDAC4, ELAVL1, SENP1, RAC2, UBD, RAC3, CDC42,...</td>\n",
       "      <td>[RAC2, CDC42, HDAC4]</td>\n",
       "      <td>3</td>\n",
       "      <td>[MSTN, ARPC2, AVIL, WNT1, RAC2, PIK3R1, PIK3CA...</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>Module_33</td>\n",
       "      <td>GO:0016567</td>\n",
       "      <td>1.433593e-03</td>\n",
       "      <td>3.902231</td>\n",
       "      <td>[HDAC4, ELAVL1, SENP1, RAC2, UBD, RAC3, CDC42,...</td>\n",
       "      <td>[UBD, ARRB2, CDC42]</td>\n",
       "      <td>3</td>\n",
       "      <td>[TMEM129, UBA6, ASB14, RFPL4A, MARCHF11, ZSWIM...</td>\n",
       "      <td>522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>Module_33</td>\n",
       "      <td>GO:0030865</td>\n",
       "      <td>9.091819e-08</td>\n",
       "      <td>8.772185</td>\n",
       "      <td>[HDAC4, ELAVL1, SENP1, RAC2, UBD, RAC3, CDC42,...</td>\n",
       "      <td>[RAC3, RHOD, RAC2]</td>\n",
       "      <td>3</td>\n",
       "      <td>[WDR1, EPB41L3, RHOH, RHOG, RHOF, U3KQA9, RHOA...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>Module_33</td>\n",
       "      <td>GO:0032956</td>\n",
       "      <td>5.968898e-06</td>\n",
       "      <td>6.686807</td>\n",
       "      <td>[HDAC4, ELAVL1, SENP1, RAC2, UBD, RAC3, CDC42,...</td>\n",
       "      <td>[RAC2, RAC3, RHOD]</td>\n",
       "      <td>3</td>\n",
       "      <td>[ODAM, SMIM22, ROCK2, HIP1R, OR2A4, FAM107A, A...</td>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>282 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Community_name       GO ID       p-value  Erichment_Score  \\\n",
       "0         Module_2  GO:0000122  2.198714e-03         3.785060   \n",
       "1         Module_2  GO:0045087  5.068671e-04         4.566379   \n",
       "2         Module_2  GO:0045944  4.076125e-03         3.441766   \n",
       "3         Module_3  GO:0000122  3.705554e-03         3.443890   \n",
       "4         Module_3  GO:0006357  1.633251e-03         3.305400   \n",
       "..             ...         ...           ...              ...   \n",
       "277      Module_33  GO:0008360  3.730783e-05         5.776614   \n",
       "278      Module_33  GO:0010592  2.493232e-07         8.263737   \n",
       "279      Module_33  GO:0016567  1.433593e-03         3.902231   \n",
       "280      Module_33  GO:0030865  9.091819e-08         8.772185   \n",
       "281      Module_33  GO:0032956  5.968898e-06         6.686807   \n",
       "\n",
       "                                       Genes in Module  \\\n",
       "0             [DISC1, SFPQ, REL, UBE2D3, GPS2, TRIM68]   \n",
       "1             [DISC1, SFPQ, REL, UBE2D3, GPS2, TRIM68]   \n",
       "2             [DISC1, SFPQ, REL, UBE2D3, GPS2, TRIM68]   \n",
       "3    [ZMYND10, CALM1, UBE2I, HIC1, TCF4, OGT, NEUROD1]   \n",
       "4    [ZMYND10, CALM1, UBE2I, HIC1, TCF4, OGT, NEUROD1]   \n",
       "..                                                 ...   \n",
       "277  [HDAC4, ELAVL1, SENP1, RAC2, UBD, RAC3, CDC42,...   \n",
       "278  [HDAC4, ELAVL1, SENP1, RAC2, UBD, RAC3, CDC42,...   \n",
       "279  [HDAC4, ELAVL1, SENP1, RAC2, UBD, RAC3, CDC42,...   \n",
       "280  [HDAC4, ELAVL1, SENP1, RAC2, UBD, RAC3, CDC42,...   \n",
       "281  [HDAC4, ELAVL1, SENP1, RAC2, UBD, RAC3, CDC42,...   \n",
       "\n",
       "             Intersecting Genes The number of intersecting genes  \\\n",
       "0          [SFPQ, UBE2D3, GPS2]                                3   \n",
       "1           [SFPQ, TRIM68, REL]                                3   \n",
       "2             [SFPQ, REL, GPS2]                                3   \n",
       "3            [OGT, UBE2I, HIC1]                                3   \n",
       "4    [OGT, NEUROD1, TCF4, HIC1]                                4   \n",
       "..                          ...                              ...   \n",
       "277          [RAC2, RAC3, RHOD]                                3   \n",
       "278        [RAC2, CDC42, HDAC4]                                3   \n",
       "279         [UBD, ARRB2, CDC42]                                3   \n",
       "280          [RAC3, RHOD, RAC2]                                3   \n",
       "281          [RAC2, RAC3, RHOD]                                3   \n",
       "\n",
       "                                          Process_Gene  \\\n",
       "0    [LINC-PINT, E2F8, FEZF1, CNOT1, NUPR2, HELT, N...   \n",
       "1    [TRGV2, DEFB131B, TRGV5, TRGV8, TRGV4, TRDV1, ...   \n",
       "2    [E2F8, ZBTB7C, NKX2-6, ZNF840P, HELT, MSGN1, N...   \n",
       "3    [LINC-PINT, E2F8, FEZF1, CNOT1, NUPR2, HELT, N...   \n",
       "4    [MT-RNR1, EPOP, UNCX, ZNF98, ZNF716, ZNF724, Z...   \n",
       "..                                                 ...   \n",
       "277  [PALM3, PALM2AKAP2, CCL24, LST1, CDC42EP2, SLC...   \n",
       "278  [MSTN, ARPC2, AVIL, WNT1, RAC2, PIK3R1, PIK3CA...   \n",
       "279  [TMEM129, UBA6, ASB14, RFPL4A, MARCHF11, ZSWIM...   \n",
       "280  [WDR1, EPB41L3, RHOH, RHOG, RHOF, U3KQA9, RHOA...   \n",
       "281  [ODAM, SMIM22, ROCK2, HIP1R, OR2A4, FAM107A, A...   \n",
       "\n",
       "    The number of components of prior_knowledge  \n",
       "0                                           968  \n",
       "1                                           585  \n",
       "2                                          1200  \n",
       "3                                           968  \n",
       "4                                          1697  \n",
       "..                                          ...  \n",
       "277                                         151  \n",
       "278                                          29  \n",
       "279                                         522  \n",
       "280                                          21  \n",
       "281                                          82  \n",
       "\n",
       "[282 rows x 9 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "returned_all "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "studied-gentleman",
   "metadata": {},
   "outputs": [],
   "source": [
    "returned_all.to_csv(f'../Outputs/AndrogenReceptor_inferred_csn_overrepresentation_results.tab',sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ambient-harris",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
